File: android/src/main/java/expo/modules/yandexmapkit/ExpoYandexMapKitPolylineView.kt
package expo.modules.yandexmapkit

import android.content.Context
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.geometry.Polyline
import com.yandex.mapkit.map.LineStyle
import com.yandex.mapkit.map.MapObject
import com.yandex.mapkit.map.MapObjectCollection
import com.yandex.mapkit.map.MapObjectTapListener
import com.yandex.mapkit.map.PolylineMapObject
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.lang.ref.WeakReference

// One `<Polyline>`. An (invisible) ExpoView React mounts as a child of the map view; it drives a
// MapKit PolylineMapObject. Because a polyline can only be created with its geometry, the object is
// created lazily once both the map collection and at least two points are available — mirroring how
// the map hands markers their placemark.
class ExpoYandexMapKitPolylineView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext), MapObjectTapListener, MapObjectChild {
  private val onPress by EventDispatcher<Map<String, Any?>>()

  private var collection: MapObjectCollection? = null
  private var mapObject: PolylineMapObject? = null
  private var points: List<Point> = emptyList()
  private var strokeColor: Int? = null
  private var strokeWidth: Float? = null
  private var outlineColor: Int? = null
  private var outlineWidth: Float? = null
  private var dashLength: Float? = null
  private var dashOffset: Float? = null
  private var gapLength: Float? = null
  private var zIndexValue = 0f
  private var handled = false

  internal fun setPoints(value: List<Point>) {
    points = value
    val obj = mapObject
    if (obj != null && obj.isValid && value.size >= 2) {
      obj.geometry = Polyline(value)
    } else {
      createIfReady()
    }
  }

  internal fun setStrokeColor(value: Int?) {
    strokeColor = value
    updateStyle()
  }

  internal fun setStrokeWidth(value: Float) {
    strokeWidth = value
    updateStyle()
  }

  internal fun setOutlineColor(value: Int?) {
    outlineColor = value
    updateStyle()
  }

  internal fun setOutlineWidth(value: Float) {
    outlineWidth = value
    updateStyle()
  }

  internal fun setDashLength(value: Float) {
    dashLength = value
    updateStyle()
  }

  internal fun setDashOffset(value: Float) {
    dashOffset = value
    updateStyle()
  }

  internal fun setGapLength(value: Float) {
    gapLength = value
    updateStyle()
  }

  internal fun setZIndexValue(value: Float) {
    zIndexValue = value
    updateStyle()
  }

  internal fun setHandled(value: Boolean) {
    handled = value
  }

  override fun attachToMap(collection: MapObjectCollection) {
    this.collection = collection
    createIfReady()
  }

  override fun detachFromMap(collection: MapObjectCollection) {
    mapObject?.let { collection.remove(it) }
    mapObject = null
    this.collection = null
  }

  // Attached once the map has handed over its collection, even if the geometry arrives later.
  override val isAttachedToMap: Boolean
    get() = collection != null

  private fun createIfReady() {
    if (mapObject != null) {
      return
    }
    val collection = collection ?: return
    if (points.size < 2) {
      return
    }
    val obj = collection.addPolyline(Polyline(points))
    mapObject = obj
    obj.addTapListener(WeakReference(this))
    updateStyle()
  }

  private fun updateStyle() {
    val obj = mapObject ?: return
    if (!obj.isValid) {
      return
    }
    obj.zIndex = zIndexValue
    obj.geometry = Polyline(points)
    strokeColor?.let { obj.setStrokeColor(it) }
    // Width / dash / outline are configured through a LineStyle (not direct properties).
    val style = LineStyle()
    strokeWidth?.let { style.strokeWidth = it }
    dashLength?.let { style.dashLength = it }
    gapLength?.let { style.gapLength = it }
    dashOffset?.let { style.dashOffset = it }
    outlineColor?.let { style.outlineColor = it }
    outlineWidth?.let { style.outlineWidth = it }
    obj.style = style
  }

  override fun onMapObjectTap(mapObject: MapObject, point: Point): Boolean {
    onPress(mapOf("point" to mapOf("latitude" to point.latitude, "longitude" to point.longitude)))
    return handled
  }
}

File: ios/ExpoYandexMapKitPolylineView.swift
import ExpoModulesCore
import YandexMapsMobile

// One `<Polyline>`. An (invisible) ExpoView React mounts as a child of the map view; it drives a
// MapKit YMKPolylineMapObject. A polyline can only be created with its geometry, so the object is
// created lazily once both the map collection and at least two points are available.
class ExpoYandexMapKitPolylineView: ExpoView, MapObjectChild {
  let onShapePress = EventDispatcher()

  private var collection: YMKMapObjectCollection?
  private var mapObject: YMKPolylineMapObject?
  private var tapListener: PolylineTapListener?
  private var points: [YMKPoint] = []
  private var strokeColor: UIColor?
  private var strokeWidth: Float?
  private var outlineColor: UIColor?
  private var outlineWidth: Float?
  private var dashLength: Float?
  private var dashOffset: Float?
  private var gapLength: Float?
  private var zIndexValue: Float = 0
  private var handled = false

  func setPoints(_ value: [PointRecord]) {
    points = value.map { YMKPoint(latitude: $0.latitude, longitude: $0.longitude) }
    if let obj = mapObject, obj.isValid, points.count >= 2 {
      obj.geometry = YMKPolyline(points: points)
    } else {
      createIfReady()
    }
  }

  func setStrokeColor(_ value: UIColor?) {
    strokeColor = value
    updateStyle()
  }

  func setStrokeWidth(_ value: Double) {
    strokeWidth = Float(value)
    updateStyle()
  }

  func setOutlineColor(_ value: UIColor?) {
    outlineColor = value
    updateStyle()
  }

  func setOutlineWidth(_ value: Double) {
    outlineWidth = Float(value)
    updateStyle()
  }

  func setDashLength(_ value: Double) {
    dashLength = Float(value)
    updateStyle()
  }

  func setDashOffset(_ value: Double) {
    dashOffset = Float(value)
    updateStyle()
  }

  func setGapLength(_ value: Double) {
    gapLength = Float(value)
    updateStyle()
  }

  func setZIndexValue(_ value: Double) {
    zIndexValue = Float(value)
    updateStyle()
  }

  func setHandled(_ value: Bool) {
    handled = value
  }

  func attachToMap(_ collection: YMKMapObjectCollection) {
    self.collection = collection
    createIfReady()
  }

  func detachFromMap(_ collection: YMKMapObjectCollection) {
    if let obj = mapObject {
      collection.remove(with: obj)
    }
    mapObject = nil
    tapListener = nil
    self.collection = nil
  }

  var isAttachedToMap: Bool { collection != nil }

  private func createIfReady() {
    guard mapObject == nil, let collection = collection, points.count >= 2 else {
      return
    }
    let obj = collection.addPolyline(with: YMKPolyline(points: points))
    mapObject = obj
    let listener = PolylineTapListener(view: self)
    tapListener = listener
    obj.addTapListener(with: listener)
    updateStyle()
  }

  private func updateStyle() {
    guard let obj = mapObject, obj.isValid else {
      return
    }
    obj.zIndex = zIndexValue
    // On iOS width / dash / outline are direct properties of the polyline object (there is no
    // setStyle method); colors go through setStrokeColorWith / the outlineColor property.
    if let strokeColor = strokeColor {
      obj.setStrokeColorWith(strokeColor)
    }
    if let strokeWidth = strokeWidth {
      obj.strokeWidth = strokeWidth
    }
    if let dashLength = dashLength {
      obj.dashLength = dashLength
    }
    if let gapLength = gapLength {
      obj.gapLength = gapLength
    }
    if let dashOffset = dashOffset {
      obj.dashOffset = dashOffset
    }
    if let outlineWidth = outlineWidth {
      obj.outlineWidth = outlineWidth
    }
    if let outlineColor = outlineColor {
      obj.outlineColor = outlineColor
    }
  }

  fileprivate func handleTap(_ point: YMKPoint) -> Bool {
    onShapePress(["point": ["latitude": point.latitude, "longitude": point.longitude]])
    return handled
  }
}

private final class PolylineTapListener: NSObject, YMKMapObjectTapListener {
  private weak var view: ExpoYandexMapKitPolylineView?

  init(view: ExpoYandexMapKitPolylineView) {
    self.view = view
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    return view?.handleTap(point) ?? false
  }
}
